### PR TITLE
[IOTDB-2112] Do not use .tsfile when compacting the target file

### PR DESCRIPTION
**Pull request by choubenson**
When processing inner space compaction, we use xxx.target file as the temporary target file and finally move it to xxx.tsfile. Also need to modify the innerSpaceRecoverCompactionTask. This change is to avoid conflict when restarting.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java [hunk 1]
```diff
@@ -584,4 +587,40 @@ public static ICrossSpaceMergeFileSelector getCrossSpaceFileSelector(
...
+    File newFile = new File(newFilePath);
+    FSFactoryProducer.getFSFactory().moveFile(oldFile, newFile);
+
+    // move .resource file
+    targetResource.setFile(newFile);
+    targetResource.setClosed(true);
+    targetResource.serialize();
+
+    // delete old tsfile and .resource file
+    File oldResourceFile = new File(oldFile.getPath() + TsFileResource.RESOURCE_SUFFIX);
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java:623]: Old tsfile is not deleted here

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java [hunk 2]
```diff
@@ -584,4 +587,40 @@ public static ICrossSpaceMergeFileSelector getCrossSpaceFileSelector(
...
+    oldResourceFile.delete();
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java:624]: check the return value of `delete`

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java [hunk 3]
```diff
@@ -65,13 +67,21 @@ public SizeTieredCompactionRecoverTask(
...
+   * targetFile is xxx.tsfile, targetResourceFile is xxx.tsfile.resource. To clear unfinished
+   * compaction task, there are several situations:
    *
    * <ol>
-   *   <li><b>Target file is uncompleted</b>: delete the target file and compaction log.
-   *   <li><b>Target file is completed, not all source files have been deleted</b>: delete the
-   *       source files and compaction logs
-   *   <li><b>Target file is completed, all source files have been deleted, compaction log file
+   *   <li>Compaction log is incomplete, then delete it and return.
+   *   <li>TmpTargetFile exists, then delete targetFile if exists and remove tmpTargetFile to
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:76]: "remove sth to sth" is not a legal English grammar. What if tmpTargetFile is not complete, we should remove TmpTargetFile and its resource then,

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java [hunk 4]
```diff
@@ -65,13 +67,21 @@ public SizeTieredCompactionRecoverTask(
...
+   * We support tmpTargetFile is xxx.target, tmpTargetResourceFile is xxx.target.resource,
+   * targetFile is xxx.tsfile, targetResourceFile is xxx.tsfile.resource. To clear unfinished
+   * compaction task, there are several situations:
    *
    * <ol>
-   *   <li><b>Target file is uncompleted</b>: delete the target file and compaction log.
-   *   <li><b>Target file is completed, not all source files have been deleted</b>: delete the
-   *       source files and compaction logs
-   *   <li><b>Target file is completed, all source files have been deleted, compaction log file
+   *   <li>Compaction log is incomplete, then delete it and return.
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:75]: I guess what you mean is "Compaction log is empty"

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java [hunk 5]
```diff
@@ -166,6 +166,24 @@ public String getFilename() {
     return filename;
   }
 
+  public String getFilePath() {
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java:169]: Only path in first data dirs will be return. Using `getFileFromDataDirs().getPath()` will be better.

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java [hunk 6]
```diff
@@ -93,33 +103,76 @@ public void doCompaction() {
...
+                        TsFileNameGenerator.COMPACTION_TMP_FILE_SUFFIX,
+                        TsFileConstant.TSFILE_SUFFIX));
+        // xxx.target.resource
+        File tmpTargetResourceFile =
+            new File(targetFileIdentifier.getFilePath() + TsFileResource.RESOURCE_SUFFIX);
+        // xxx.tsfile.resource
+        File targetResourceFile = new File(targetFile.getPath() + TsFileResource.RESOURCE_SUFFIX);
+
+        if (tmpTargetFile != null) {
+          // xxx.target exists, then remove it to xxx.tsfile
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:132]: move it

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java [hunk 7]
```diff
@@ -93,33 +103,76 @@ public void doCompaction() {
...
+          if (targetFile.exists()) {
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:133]: What if the tmpTargetFile is not complete?

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java (hunk 5 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileIdentifier.java:169]: But if targetFile does  not exist, getFileFromDataDirs() will return null.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileNameGenerator.java [hunk 8]
```diff
@@ -38,6 +38,7 @@
 public class TsFileNameGenerator {
 
   private static FSFactory fsFactory = FSFactoryProducer.getFSFactory();
+  public static final String COMPACTION_TMP_FILE_SUFFIX = ".target";
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileNameGenerator.java:41]: move to IoTDBConstant

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java [hunk 9]
```diff
@@ -584,4 +587,46 @@ public static ICrossSpaceMergeFileSelector getCrossSpaceFileSelector(
...
+   * xxx.tsfile.resource.
+   *
+   * @param targetResource the old tsfile to be moved, which is xxx.target
+   * @throws IOException
+   */
+  public static void moveTargetFile(TsFileResource targetResource, String fullStorageGroupName)
+      throws IOException {
+    if (!targetResource.getTsFilePath().endsWith(TsFileNameGenerator.COMPACTION_TMP_FILE_SUFFIX)) {
+      logger.warn(
+          "TmpTargetFileReource "
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java:602]:
```suggestion
          "Tmp target TsFile "
```

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java [hunk 10]
```diff
@@ -584,4 +587,46 @@ public static ICrossSpaceMergeFileSelector getCrossSpaceFileSelector(
...
+    targetResource.setClosed(true);
+    targetResource.serialize();
+
+    // delete old resource file
+    File oldResourceFile = new File(oldFile.getPath() + TsFileResource.RESOURCE_SUFFIX);
+    if (!oldResourceFile.delete()) {
+      logger.error(
+          "{} [Compaction] Fail to delete old resource file {}",
+          fullStorageGroupName,
+          oldResourceFile.getAbsolutePath());
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java:629]: Use FSFactoryProducer.getFSFactory().moveFile(oldFile, newFile) to move resource, this is a atomic action.

Do not use write new, then delete. You will meet two .resource when restarting.

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java [hunk 11]
```diff
@@ -105,6 +105,7 @@ protected void doCompaction() throws Exception {
     long startTime = System.currentTimeMillis();
     // get resource of target file
     String dataDirectory = selectedTsFileResourceList.get(0).getTsFile().getParent();
+    // Here is tmpTargetFile, which is xxx.target
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionTask.java:108]: will this .target be conflict with merge target file?

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java [hunk 12]
```diff
@@ -93,33 +104,88 @@ public void doCompaction() {
...
+            getFileFromDataDirs(
+                targetFileIdentifier
+                    .getFilePath()
+                    .replace(
+                        TsFileNameGenerator.COMPACTION_TMP_FILE_SUFFIX,
+                        TsFileConstant.TSFILE_SUFFIX));
+        // xxx.target.resource
+        File tmpTargetResourceFile =
+            getFileFromDataDirs(
+                targetFileIdentifier.getFilePath() + TsFileResource.RESOURCE_SUFFIX);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:129]:
```suggestion
            new File(tmpTargetFile.getPath()+ TsFileResource.RESOURCE_SUFFIX);
```

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java [hunk 13]
```diff
@@ -93,33 +104,88 @@ public void doCompaction() {
...
+
+        // xxx.tsfile.resource
+        File targetResourceFile =
+            getFileFromDataDirs(
+                targetFileIdentifier
+                        .getFilePath()
+                        .replace(
+                            TsFileNameGenerator.COMPACTION_TMP_FILE_SUFFIX,
+                            TsFileConstant.TSFILE_SUFFIX)
+                    + TsFileResource.RESOURCE_SUFFIX);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:139]:
```suggestion
            new File(targetFile.getPath()+ TsFileResource.RESOURCE_SUFFIX);
```

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java [hunk 14]
```diff
@@ -93,33 +104,88 @@ public void doCompaction() {
...
+
+        if (tmpTargetFile != null) {
+          // xxx.target exists, then move it to xxx.tsfile
+          if (targetFile != null) {
+            targetFile.delete();
+          }
+          TsFileResource targetResource = new TsFileResource(tmpTargetFile);
+          InnerSpaceCompactionUtils.moveTargetFile(targetResource, fullStorageGroupName);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:147]: why fullStorageGroupName?

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java [hunk 15]
```diff
@@ -94,93 +99,78 @@ public void doCompaction() {
...
-          }
-
-          InnerSpaceCompactionUtils.deleteTsFilesInDisk(
-              sourceTsFileResources, fullStorageGroupName);
-          InnerSpaceCompactionUtils.deleteModificationForSourceFile(
-              sourceTsFileResources, logicalStorageGroupName + "-" + virtualStorageGroup);
+          handleSuccess =
+              InnerSpaceCompactionExceptionHandler.handleWhenAllSourceFilesExist(
+                  fullStorageGroupName, targetResource, sourceResources);
+          //  handleWithAllSourceFilesExist(targetFileIdentifier, sourceFileIdentifiers);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/sizetiered/SizeTieredCompactionRecoverTask.java:145]:
```suggestion
```

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**